Animation rules in normal restyle

hiikezoe opened the pull request:
<!-- Please describe your changes on the following line: -->

This is a PR for https://bugzilla.mozilla.org/show_bug.cgi?id=1366631 and https://bugzilla.mozilla.org/show_bug.cgi?id=1367225

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because it's for stylo

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17014)
<!-- Reviewable:end -->

Commit:
Use SMIL override value that has been processed during animation-only restyles for normal restyle.

In the case where we process an element which has SMIL animations in normal travesal
the SMIL styles must have been computed in animation-only restyles. So we
have only to pick the computed styles instead of recomputing it.

Commit:
Move AnimationRules into declaration_block.rs.

Commit:
Use animation values that have been processed during animation-only restyle for normal restyle.

Commit:
Check ElementHasAnimations before trying to get animations rules.

In case of ::before and ::after element, the generated content has no
ElementHasAnimations flag, their parent has the flag.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/data.rs, components/style/rule_tree/mod.rs, components/style/properties/declaration_block.rs, components/style/dom.rs, components/style/gecko/wrapper.rs and 2 more
 * @emilio: components/style/data.rs, components/style/rule_tree/mod.rs, components/style/properties/declaration_block.rs, components/style/dom.rs, components/style/gecko/wrapper.rs and 2 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

hiikezoe commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit fce7c2d has been approved by `emilio`

<!-- @bors-servo r=emilio fce7c2d885ee326e750591818a4a7c9ad7070aff -->

hiikezoe commented:
I probably made a mistake during rebasing.
@bors-servo r-

hiikezoe commented:
oops, I saw a different try result..

Commit:
Make replace_rules returning boolean.

We only use whether the return value is IMPORTANT_RULES_CHANGED or not, so we
can just return true if an important rules was changed in the function.
Also, we can just return false in case of animation rules changes sine for
animation we can ensure there is no importan rules.
Because of these changes, replace_rule_node does not borrow |result| so that
we can drop a scope there.

Commit:
Don't process style attribute changes in animation-only restyle.

hiikezoe commented:
@bors-servo r=emilio,birtles

bors-servo commented:
:pushpin: Commit 976b3ba has been approved by `emilio,birtles`

<!-- @bors-servo r=emilio,birtles 976b3ba6e79b4c2682b961a5defe88f5e1195d00 -->
